include/ada/url-inl.h

```diff
@@ -73,6 +79,107 @@ namespace ada {
+  }
+
+  // faster than std::to_string(x).size().
+  inline int fast_digit_count(uint32_t x) {
+    // Compiles to very few instructions. Note that the
+    // table is static and thus effectively a constant.
+    // We leave it inside the function because it is meaningless
+    // outside of it (this comes at no performance cost).
+    const static uint64_t table[] = {
+      4294967296,  8589934582,  8589934582,  8589934582,  12884901788,
+      12884901788, 12884901788, 17179868184, 17179868184, 17179868184,
+      21474826480, 21474826480, 21474826480, 21474826480, 25769703776,
+      25769703776, 25769703776, 30063771072, 30063771072, 30063771072,
+      34349738368, 34349738368, 34349738368, 34349738368, 38554705664,
+      38554705664, 38554705664, 41949672960, 41949672960, 41949672960,
+      42949672960, 42949672960};
+    return int((x + table[int_log2(x)]) >> 32);
+  }
+
+  [[nodiscard]] ada_really_inline ada::url_components url::get_components() noexcept {
+    url_components out{};
+
+    // protocol ends with ':'. for example: "https:"
+    out.protocol_end = uint32_t(get_scheme().size());
+
+    // Trailing index is always the next character of the current one.
+    size_t trailing_index = out.protocol_end + 1;
+
+    if (host.has_value()) {
+      // 2 characters for "//" and 1 character for starting index
+      out.host_start = out.protocol_end + 3;
+
+      if (includes_credentials()) {
+        out.username_end = uint32_t(out.host_start + username.size() - 1);
+
+        out.host_start += uint32_t(username.size() + 1);
+
+        if (!password.empty()) {
+          out.host_start += uint32_t(password.size() + 1);
+        }
+      } else {
+        out.username_end = out.host_start;
+      }
+
+      out.host_end = uint32_t(out.host_start + host.value().size()) - 1;
+      trailing_index = out.host_end + 1;
+    } else {
+      out.host_start = out.protocol_end + 1;
+      out.host_end = out.protocol_end + 1;
+
+      size_t url_delimiter_count = std::count(path.begin(), path.end(), '/');
+
+      if (!has_opaque_path && url_delimiter_count > 1 && path.length() >= 2 && path[0] == '/' && path[1] == '/') {
+        // If url’s host is null, url does not have an opaque path, url’s path’s size is greater than 1,
+        // and url’s path[0] is the empty string, then append U+002F (/) followed by U+002E (.) to output.
+        trailing_index = out.protocol_end + 3;
+      } else {
+        trailing_index = out.protocol_end + 1;
+      }
+    }
+
+    if (port.has_value()) {
+      out.port = port.value();
+      trailing_index += fast_digit_count(port.value()) + 1; // Port omits ':'
+    }
+
+    if (!path.empty()) {
+      out.pathname_start = uint32_t(trailing_index);
+      trailing_index += path.size();
+    } else {
+      out.pathname_start = uint32_t(trailing_index);
+    }
+
+    if (query.has_value()) {
+      out.search_start = uint32_t(trailing_index);
+      trailing_index += get_search().size();
```

Comment:
trailing_index is fine. Another name for the same concept might be 'running_index'.